### ISSUE TITLE
The ability to reference data in configurations in tasks.json

### ISSUE THREAD

**Issue by LouChiSoft** [microsoft/vscode-cpptools#7456]
Type: Feature Request

A method of being able to reference a C/C++ configuration you have created in the tasks.json file to enable the passing of information and reduce the amount of work needed to keep tasks and configs the same.

Example usage:
```
{
  "version": "2.0.0",
  "tasks": [
    {
      "type": "shell",
      "label": "C/C++: g++.exe build all files",
      "configuration": "GCC"      <--- Line is needed when referencing configuration data and must match the name of a configuration
      "command": "C:\\Program Files\\mingw-w64\\x86_64-8.1.0-posix-seh-rt_v6-rev0\\mingw64\\bin\\g++.exe",
      "args": [
        "-g", 
        "${workspaceFolder}/*.cpp",
        "-L${configurationLibraryDirectory}",  <-- Functions as a wildcard for the library directories
        "-l${configurationsLibraries},         <-- Functions as a wildcard for the libraries
        "-D${configurationDefines}",           <-- Functions as a wildcard for the compiler defines
        "-o", 
        "${workspaceFolderf}/out.exe"],
      "options": {
        "cwd": "${workspaceFolder}"
      },
      "problemMatcher": ["$gcc"],
      "group": {
        "kind": "build",
        "isDefault": true
      }
    }
  ]
}
```
The intention here would allow the user to add information straight from the configuration and pass that into the build command. There will be an error/warning if you try to reference a configuration without a name and any tasks that the user runs without adding the configuration name will ignore all configuration args. Additionally upon calling a build task that has a configuration referenced IntelliSense should automatically swap to that configuration if it isn't already selected. This should overall save time in maintaining the IntelliSense file and tasks file by only needing to change one file.

**Comment by sean-mcmanus** [comment]
I think the "type" would need to be "cppBuild" and not "shell". Compiler defines shouldn't need to be added to a build task...I assume you mean other defines not defined by the compiler.

However, I'm not sure we'd want to implement this, because all the existing designs have users set the build and IntelliSense may auto-consume it because usually users have their build command determined beforehand as a higher priority than IntelliSense configuration, but this is doing it the other way around...however, this seems like it'd be an opt-in feature so maybe that would be okay.

**Comment by LouChiShu** [comment]
Yes, sorry compiler defines was the incorrect term to use, I did mean additional defines defined within the config. I do agree that the actual building of the project would be a higher priority than the IntelliSense, but as you already pointed out this would be an opt-in feature to those who would like to operate IntelliSense first. Additionally I can't think of a method of consuming data from the build task within the configuration that would be quite as simple as the one I have suggested.

Do you mind if I ask what the difference between "shell" and "cppBuild" is? My example was based upon the getting started page on Microsoft's website and I haven't come across "cppBuild" before

**Comment by sean-mcmanus** [comment]
We added "cppBuild" task types a few months ago...seems the getting started hasn't been updated yet. The main disadvantage with cppBuild tasks is probably that the no colored fonts are used. We generate a cppBuild task in certain cases such as the Build and Debug Active file command...it gives us more control over the task than the shell task does.
